### ISSUE TITLE
local_action requires "localhost" in inventory Or URI module throws httplib2 error.

### ISSUE THREAD

**Issue by haridas** [ansible/ansible#14953]
##### Issue Type:

- Bug Report
##### Ansible Version:

$ ansible --version
ansible 2.0.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
##### Ansible Configuration:

Working on default settings for this ansible version. No changes made.
##### Environment:

Control machine -> Ubuntu 15.10
Remote Host - ubuntu/trusty64 on vagrant
##### Summary:

Via `local_action` I'm calling a API using `uri` module, If `localhost` wasn't present in the inventory file then I'm getting this error `"msg": "httplib2 >= 0.7 is not installed"`. If I add the `localhost` in inventory under any group, it works without any issue. The behavior is `localhost` should be implicitly used for `local_action` right ?
##### Steps To Reproduce:

I'm querying a remote API, and based on that doing some updates on the remote host, ( Vagrant is the remote machine). The hosts file looks as 

_command_
`$ ansible-playbook  -i hosts1 playbook.yml --limit bms`

_hosts_

```
# inventory 1 - hosts file which works
[local]
localhost

[bms]
bms1 ansible_host=localhost ansible_user=vagrant ansible_port=2226

```

_playbook.yml_

```
- hosts: bms
  any_errors_fatal: true
  serial: 1


  pre_tasks:
    - name: Request to release "{{ inventory_hostname }}".
      local_action: uri url="{{ release_bms_url }}?bms_ips={{inventory_hostname}}" return_content=yes status_code=200
      register: result
      until: result.json is defined and result.json.get("{{inventory_hostname}}")
      retries: "{{ bms_release_retry }}"
      delay: "{{ bms_release_delay }}"

    - name: Echo
      debug: msg={{ result }}

    - fail: msg="Failed to Release BMS:{{ inventory_hostname }}"
      when: result.json is defined and not result.json.get("{{inventory_hostname}}")

```

_groups_vars/bms_

```
release_bms_url: http://localhost:8081/admin/releaseBms
# Time in seconds.
bms_release_delay: 1
# Retries this many times and fails after that.
bms_release_retry: 3
```

With the above setup and command, it works correctly as specified way. But problem comes as in the host file the reference to the `localhost` is there, and I read that `local_action` implicitly refer the localhost. So I removed that from the hosts file, so the new hosts file looks like bellow

_hosts1_

```
# inventory 2 -> The hosts file which doesn't work.
# The playbook will fail to run by throwing httplib2 missing error.
[bms]
bms1 ansible_host=localhost ansible_user=vagrant ansible_port=2226
```

If I remove the `localhost` from the hosts file, as given above for `hosts1`, I'm getting this `"msg": "httplib2 >= 0.7 is not installed"` Error.
##### Expected Results:

`local_action` should pick the reference to `localhost` implicitly. I can see some commits and merges happened for this issue, but it's not working with this version.
##### Running the play book without localhost in the inventory [ Throws Error ]

```

(ENV)haridas@IMCHLT015:~/projects$ ansible-playbook  -i deploy/ansible/hosts1 deploy/ansible/test-play.yml --limit bms -vvvv
Using /etc/ansible/ansible.cfg as config file
Loaded callback default of type stdout, v2.0
1 plays in deploy/ansible/test-play.yml

PLAY ***************************************************************************

TASK [setup] *******************************************************************
<localhost> ESTABLISH SSH CONNECTION FOR USER: vagrant
<localhost> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=2226 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=10 -o ControlPath=/home/haridas/.ansible/cp/ansible-ssh-%h-%p-%r -tt localhost '/bin/sh -c '"'"'( umask 22 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1457945894.25-153266588796827 `" && echo "` echo $HOME/.ansible/tmp/ansible-tmp-1457945894.25-153266588796827 `" )'"'"''
<localhost> PUT /tmp/tmpmNKiYn TO /home/vagrant/.ansible/tmp/ansible-tmp-1457945894.25-153266588796827/setup
<localhost> SSH: EXEC sftp -b - -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=2226 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=10 -o ControlPath=/home/haridas/.ansible/cp/ansible-ssh-%h-%p-%r '[localhost]'
<localhost> ESTABLISH SSH CONNECTION FOR USER: vagrant
<localhost> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=2226 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=10 -o ControlPath=/home/haridas/.ansible/cp/ansible-ssh-%h-%p-%r -tt localhost '/bin/sh -c '"'"'LANG=en_IN LC_ALL=en_IN LC_MESSAGES=en_IN /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1457945894.25-153266588796827/setup; rm -rf "/home/vagrant/.ansible/tmp/ansible-tmp-1457945894.25-153266588796827/" > /dev/null 2>&1'"'"''
ok: [bms1]

TASK [Request to release "bms1".] **********************************************
task path: /mnt/data/projects/deploy/ansible/test-play.yml:8
ESTABLISH LOCAL CONNECTION FOR USER: haridas
localhost EXEC /bin/sh -c '( umask 22 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1457945895.99-202072692879422 `" && echo "` echo $HOME/.ansible/tmp/ansible-tmp-1457945895.99-202072692879422 `" )'
localhost PUT /tmp/tmp8p98Bn TO /home/haridas/.ansible/tmp/ansible-tmp-1457945895.99-202072692879422/uri
localhost EXEC /bin/sh -c 'LANG=en_IN LC_ALL=en_IN LC_MESSAGES=en_IN /mnt/data/projects/ENV/bin/python /home/haridas/.ansible/tmp/ansible-tmp-1457945895.99-202072692879422/uri; rm -rf "/home/haridas/.ansible/tmp/ansible-tmp-1457945895.99-202072692879422/" > /dev/null 2>&1'
fatal: [bms1 -> localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_args": {"backup": null, "body": null, "body_format": "raw", "content": null, "creates": null, "delimiter": null, "dest": null, "directory_mode": null, "follow": false, "follow_redirects": "safe", "force": null, "force_basic_auth": false, "group": null, "method": "GET", "mode": null, "owner": null, "password": null, "regexp": null, "remote_src": null, "removes": null, "return_content": true, "selevel": null, "serole": null, "setype": null, "seuser": null, "src": null, "status_code": ["200"], "timeout": 30, "url": "http://localhost:8081/admin/releaseBms", "user": null, "validate_certs": true}, "module_name": "uri"}, "msg": "httplib2 >= 0.7 is not installed"}

NO MORE HOSTS LEFT *************************************************************
        to retry, use: --limit @deploy/ansible/test-play.retry

PLAY RECAP *********************************************************************
bms1                       : ok=1    changed=0    unreachable=0    failed=1 

```
##### Running the play book with localhost in the inventory [ Works correctly ]

```
(ENV)haridas:~/projects$ ansible-playbook  -i deploy/ansible/hosts deploy/ansible/test-play.yml --limit bms -vvvv
Using /etc/ansible/ansible.cfg as config file
Loaded callback default of type stdout, v2.0
1 plays in deploy/ansible/test-play.yml

PLAY ***************************************************************************

TASK [setup] *******************************************************************
<localhost> ESTABLISH SSH CONNECTION FOR USER: vagrant
<localhost> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=2226 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=10 -o ControlPath=/home/haridas/.ansible/cp/ansible-ssh-%h-%p-%r -tt localhost '/bin/sh -c '"'"'( umask 22 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1458365805.21-176363564848880 `" && echo "` echo $HOME/.ansible/tmp/ansible-tmp-1458365805.21-176363564848880 `" )'"'"''
<localhost> PUT /tmp/tmpEi5a8V TO /home/vagrant/.ansible/tmp/ansible-tmp-1458365805.21-176363564848880/setup
<localhost> SSH: EXEC sftp -b - -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=2226 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=10 -o ControlPath=/home/haridas/.ansible/cp/ansible-ssh-%h-%p-%r '[localhost]'
<localhost> ESTABLISH SSH CONNECTION FOR USER: vagrant
<localhost> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=2226 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=10 -o ControlPath=/home/haridas/.ansible/cp/ansible-ssh-%h-%p-%r -tt localhost '/bin/sh -c '"'"'LANG=en_IN LC_ALL=en_IN LC_MESSAGES=en_IN /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1458365805.21-176363564848880/setup; rm -rf "/home/vagrant/.ansible/tmp/ansible-tmp-1458365805.21-176363564848880/" > /dev/null 2>&1'"'"''
ok: [bms1]

TASK [Request to release "bms1".] **********************************************
task path: /mnt/data/projects/deploy/ansible/test-play.yml:13
ESTABLISH LOCAL CONNECTION FOR USER: haridas
localhost EXEC /bin/sh -c '( umask 22 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1458365807.04-214823872755448 `" && echo "` echo $HOME/.ansible/tmp/ansible-tmp-1458365807.04-214823872755448 `" )'
localhost PUT /tmp/tmpz2m0lO TO /home/haridas/.ansible/tmp/ansible-tmp-1458365807.04-214823872755448/uri
localhost EXEC /bin/sh -c 'LANG=en_IN LC_ALL=en_IN LC_MESSAGES=en_IN /usr/bin/python /home/haridas/.ansible/tmp/ansible-tmp-1458365807.04-214823872755448/uri; rm -rf "/home/haridas/.ansible/tmp/ansible-tmp-1458365807.04-214823872755448/" > /dev/null 2>&1'
ok: [bms1 -> localhost] => {"access_control_allow_headers": "Content-Type", "access_control_allow_methods": "GET,PUT,POST,DELETE", "access_control_allow_origin": "*", "changed": false, "connection": "keep-alive", "content": "{\"bms1\":true}", "content_length": "13", "content_location": "http://localhost:8081/admin/releaseBms?bms_ips=bms1", "content_type": "application/json; charset=utf-8", "date": "Sat, 19 Mar 2016 05:36:47 GMT", "etag": "W/\"d-pi+6q3wbMLFFSVVm15+xug\"", "invocation": {"module_args": {"backup": null, "body": null, "body_format": "raw", "content": null, "creates": null, "delimiter": null, "dest": null, "directory_mode": null, "follow": false, "follow_redirects": "safe", "force": null, "force_basic_auth": false, "group": null, "method": "GET", "mode": null, "owner": null, "password": null, "regexp": null, "remote_src": null, "removes": null, "return_content": true, "selevel": null, "serole": null, "setype": null, "seuser": null, "src": null, "status_code": ["200"], "timeout": 30, "url": "http://localhost:8081/admin/releaseBms?bms_ips=bms1", "user": null, "validate_certs": true}, "module_name": "uri"}, "json": {"bms1": true}, "redirected": false, "status": 200, "x_powered_by": "Express"}

TASK [fail] ********************************************************************
task path: /mnt/data/projects/deploy/ansible/test-play.yml:20
skipping: [bms1] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}

PLAY RECAP *********************************************************************
bms1                       : ok=2    changed=0    unreachable=0    failed=0
```

I can see that from `ansible-2.1` the requirement for httplib2 was removed.

**Comment by bcoca** [comment]
can you test with 'localhost' in inventory with -vvvv and show the output?

**Comment by haridas** [comment]
@bcoca Updated the issue with playbook output for inventory with `localhost` in it. Please check it.

**Comment by sivel** [comment]
From what I can see it looks like you are running Ansible out of a virtualenv called `ENV`, is this correct?

When Ansible uses the implicit localhost (not listed in inventory) it will use the Python binary in the venv (by using `sys.executable`) requiring that httplib2 be installed in the venv.

When you specify localhost in the inventory without `ansible_python_interpreter` it will utilize `/usr/bin/python` which has access to packages installed globally.  I am assuming that httplib2 is only installed for `/usr/bin/python`

**Comment by haridas** [comment]
Yes, I'm running in a virtualenv.

I checked the `/usr/bin/python`, it has `httplib2` installed. As till `ansible 2.1

**Comment by haridas** [comment]
Yes, I'm running in a virtualenv, and the ENV doesn't have `httplib2` installed. I installed it and tried to run the playbook without including `localhost` in the inventory, it works for me. From `-vvvv` I can see the uri command picked up correct `python` path from ENV.

```
localhost PUT /tmp/tmpfgaGm4 TO /home/haridas/.ansible/tmp/ansible-tmp-1458455732.27-252617888075675/uri
localhost EXEC /bin/sh -c 'LANG=en_IN LC_ALL=en_IN LC_MESSAGES=en_IN /mnt/data/projects/ENV/bin/python /home/haridas/.ansible/tmp/ansible-tmp-1458455732.27-252617888075675/uri; rm -rf "/home/haridas/.ansible/tmp/ansible-tmp-1458455732.27-252617888075675/" > /dev/null 2>&1'
```

I checked the `/usr/bin/python`, it has `httplib2` installed. As I can see, till `ansible 2.1` the `uri` module have requirement of `httplib2` on the target machine. So In my case I think, when I put `localhost` in the inventory, `local_action` consider it as explicitly supplied and using it.

When I run from the in `ENV` without `localhost` it automatically picking the `python` interpreter from `ENV`.

I think this is the expected behavior then. Thanks @sivel  for clarifying this.
